### PR TITLE
heap: merge freed extents

### PR DESCRIPTION
**Pull request by ldorau**
This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/153)

### CODE REVIEW

**Comment by kilobyte** [comment]
It's... beautiful!

I've rebased my old eyecandy visualization branch: [base](https://github.com/kilobyte/vmemcache/tree/visual), [merged](https://github.com/kilobyte/vmemcache/tree/visual-heap-merge).  Then, compile both, and run on two screens:
```
benchmarks/bench_micro /tmp/ put 4 4294967295
```

**Comment by marcinslusarz** [comment]
@kilobyte please submit your visualization code, it's too cool to be kept out of tree

File: src/vmemcache_heap.c [hunk 1]
```diff
@@ -38,15 +38,116 @@
...
+#define FLAG_ALLOCATED ((size_t)1 << (8 * HEADER_SIZE - 1))
+
+/* value of the header ('ptr' should not be in brackets) */
+#define HEADER(ptr) (*(size_t *)((uintptr_t)ptr))
+
+/* do an arithmetic on the 'void' pointer ('ptr' should not be in brackets) */
+#define VOID_PTR(ptr) ((void *)((uintptr_t)ptr))
+
+#define HEADER_VALUE(size, is_allocated) \
+	(is_allocated) ? ((size) | FLAG_ALLOCATED) : (size)
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/vmemcache_heap.c]: Done

File: src/vmemcache_heap.c [hunk 2]
```diff
@@ -38,15 +38,116 @@
...
+static inline void
+vmcache_heap_write_headers(struct heap_entry *he, int is_allocated)
+{
+	size_t header_value = HEADER_VALUE(he->size, is_allocated);
+
+	/* save the header */
+	HEADER(he->ptr) = header_value;
+
+	/* save the footer */
+	HEADER(he->ptr + he->size - HEADER_SIZE) = header_value;
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/vmemcache_heap.c]: Done

**Pull request merged**